Documentation on Pooled Staking (Theoretical) (#145)
moondance-labs/tanssi-docs: learn/tanssi/staking.md

Reviewer: This is a nitpick, but we switch between calling it Auto-Compounded Rewards Pool and Auto-Compound Rewards Pool.

---
title: Staking
description: Learn how Tanssi implements a novel Staking mechanism to provide a decentralized and trustless set of block producers to all Tanssi-powered Appchains
---

# Tanssi Staking {: #tanssi-staking }

## Introduction {: #introduction }

One of Tanssi's core propositions is to simplify the infrastructure complexity for Appchains. A significant component is bootstrapping a decentralized set of block producers, which Tanssi offers through its unique architecture and staking mechanics.

> Tanssi staking mechanics guarantee that the block producers for Tanssi and Tanssi-powered Appchains are selected through a trustless and decentralized mechanism. They also incentivize the community to delegate to top-performing or engaged block producers.

Author: I prefer the current wording because people delegate to either top-performing (in terms of rewards, I guess) or engaged block producers (like engaged community members)

This page covers the fundamental concepts of Tanssi's staking mechanics and how it secures a decentralized block production set that drives network liveness for Tanssi Appchains.

## Core Concepts {: #core-concepts }

Tanssi's staking module mechanics were inspired by the concept of liquidity pool tokens (LP tokens) in traditional Automated-Market-Makers (AMMs) like Uniswap V2.

Each block producer has four liquidity pools through which delegators move as they perform different staking operations. In short, each liquidity pool represents a different state throughout the staking process: joining, staking through manual rewards, staking through auto-compound rewards, and leaving. Nevertheless, one core difference is that LP tokens in common AMMs are transferable while staking shares tokens are not.

> A delegator has four simple transactions to go through the different states (liquidity pools): delegate (for manual or auto-compound rewards), undelegate, swap, and execute pending operations. For example, users who want to stake through either rewards pool can use the `delegate` call and join the Joining Pool immediately. After a delay, users (or anyone else) can `execute pending operation` and enter the initially set rewards pool. Once there, users can `swap` between reward pools as often as they like. Lastly, users in a rewards pool can use the `undelegate` call to go into the Leaving Pool and unstake their tokens (or anyone else's) `executing pending operation` after a given delay.

Reviewer 2: I think the use of back ticks here is a bit out of the norm. If referring to a specific function, it should be wrapped in back ticks, and the spelling should be the same as the function name, i.e., `executePendingOperations`. If you're just saying that anyone can execute the pending operations, then I wouldn't wrap it in back ticks. 

Since we're not actually defining the exact extrinsics here, I just wouldn't use back ticks

Author: sounds good

> A delegator has four simple transactions to go through the different states (liquidity pools): delegate (for manual or auto-compound rewards), undelegate, swap, and execute pending operations. For example, users who want to stake through either rewards pool can use the delegate call and join the Joining Pool immediately. After a delay, users (or anyone else) can execute the pending operation and enter the initially set rewards pool. Once there, users can swap between reward pools as often as they like. Lastly, users in a rewards pool can use the undelegate call to go into the Leaving Pool and unstake the tokens by them (or anyone else) executing the pending request after a given delay.

Reviewer: ```suggestion
A delegator has four simple transactions to go through the different states (liquidity pools): delegate (for manual or auto-compound rewards), undelegate, swap, and execute pending operations. For example, users who want to stake through either rewards pool can use the `delegate` call and join the Joining Pool immediately. After a delay, users (or anyone else) can execute the pending operation and enter the initially set rewards pool. Once there, users can swap between reward pools as often as they like. Lastly, users in a rewards pool can use the `undelegate` call to go into the Leaving Pool and unstake their tokens (or anyone else's) executing the pending request after a given delay.
```

> Liquidity pools have a set of shares that can be considered LP tokens in traditional AMMs. When users join a new liquidity pool, they are given several shares (LP tokens) that depend on the pool type, number of tokens staked, total number of shares, and total number of tokens staked in that pool.

Reviewer 2: ```suggestion
Liquidity pools have a set of shares that can be considered LP tokens in traditional AMMs. When users join a new liquidity pool, they are given several shares (LP tokens) that depend on the pool type, the number of tokens they staked, the total number of shares, and the total number of tokens staked in that pool.
```

Rewards are assigned to a block's producer Manual or Auto-Compounded Reward Pools when Tanssi attests that the specific block production slot that block producer was assigned to has been fulfilled, and the block was produced successfully.

All rewards (for all pools) are stored in a protocol-owned account. Nevertheless, the protocol internally keeps track of the actual native tokens held by each pool. The core difference between staking through the Manual or Auto-Compound Rewards Pools is how rewards are distributed. In the Manual Rewards Pool, users have to claim any staking rewards they've accumulated manually. In contrast, in the Auto-Compound Rewards Pool, the rewards are automatically re-staked at each Tanssi block, where the protocol announces the block producer for each block production assignment, either in Tanssi or an Appchain.

> The delegate and undelegate operations need to be sent by the delegator itself. They signal the intent of the action to be taken and ask the protocol to perform the necessary checks to allow the delegator to delegate or undelegate. Consequently, these actions must be executed after a certain number of sessions, but anyone in the network can perform this second operation through the execute pending operation transaction.

Reviewer: ```suggestion
The delegate and undelegate operations need to be sent by the delegator itself. They signal the intent of the action to be taken and ask the protocol to perform the necessary checks to allow the delegator to delegate or undelegate. Consequently, these actions can be executed only after a certain number of sessions, but anyone in the network can perform this second operation through the execute pending operation transaction.
```

Reviewer: Changing the wording slightly to emphasize that the pending requests can only be executed after a certain period of time, but if I'm understanding correctly, these pending requests will queue up indefinitely until someone executes them (like Moonbeam's)

The following diagram summarizes the high-level flow of a delegator delegating and undelegating tokens to a block producer. User actions are highlighted in cyan, while different pools are highlighted in coral.

![High-level overview of Tanssi Staking mechanics](/images/learn/tanssi/staking/staking-1.webp)

> ## Staking Paremeters {: #staking-parameters }

Reviewer 2: ```suggestion
## Staking Parameters {: #staking-parameters }
```

=== "Dancebox"
    |             Variable             |                                                                         Value                                                                         |
    |:--------------------------------:|:-----------------------------------------------------------------------------------------------------------------------------------------------------:|

>     |          Joining Delay          |                        {{ networks.dancebox.staking.joining_delay_blocks }} blocks ({{ networks.dancebox.staking.joining_delay_hours }} hours)                        |

Reviewer 3: Joining/Leaving hours would be 1 hour from now on

Author: You mean this is with 6 second block time? I mean sure let's update the YAML already considering that

    |          Leaving Delay          |                        {{ networks.dancebox.staking.leaving_delay_blocks }} blocks ({{ networks.dancebox.staking.leaving_delay_hours }} hours)                        |

## Staking Pools {: #staking-pools}

The following section goes through each of the liquidity pools that represent a step throughout the staking process.

### Joining Pool {: $joining-pool}

> When a user first delegates to start the staking process, it must state what staking rewards mechanism it wants: manual or auto-compound rewards (each being a separate pool). Once the joining transaction is executed, the user automatically enters the Joining Pool and is given shares of that pool directly correlated to the number of tokens being staked. This pool offers stability to the current set of block producers by providing a delay between a delegator staking and receiving rewards. The delay is set to at least one entire session.

Reviewer 2: In the above table, it says there is a delay of 600 blocks, which is two sessions. So it looks like this needs to be updated?

Author: Got catch, it should be 1200 blocks now that sessions are still set to 1 hour but we are producing blocks at 6 seconds

Similarly, when a user stakes or unstakes tokens, rewards are automatically claimed, and the user's checkpoint rewards counter is reset. Adding or removing a stake means that reward conditions for that specific amount differ from what the protocol has in storage. Consequently, the rewards counter checkpoint must be synced with the pool's rewards counter to ensure no imbalances.

![High-level overview of the Manual Rewards Pool when Staking](/images/learn/tanssi/staking/staking-3.webp)

### Auto-Compound Rewards Pool {: #autocompounded-rewards-pool}

When a user joins the Auto-Compound Rewards Pool, the protocol destroys all Joining Pool shares they own in favor of the native protocol token. Next, in the same block, the protocol computes the amount of Auto-Compound shares that can be minted with this amount based on the share's price. The price is calculated based on current pool conditions, that is, the amount of native tokens and shares that exist:

```mathematica
SharePrice [Tokens/Shares] = NumberOfTokensInPool / NumberOfSharesInPool 
```

> Shares don't have decimals. Consequently, any remaining native tokens when acquiring the pool's shares are refunded to the user. The share price is not impacted by users joining the pool, as the ratio is maintained. Once the user has Manual Rewards Pool shares, they earn staking rewards (that is, in the same session).

Reviewer 2: ```suggestion
Shares don't have decimals. Consequently, any remaining native tokens when acquiring the pool's shares are refunded to the user. The share price is not impacted by users joining the pool, as the ratio is maintained. Once the user has Auto-Compound Rewards Pool shares, they earn staking rewards (that is, in the same session).
```

### Leaving Pool {: #leaving-pool}

When a user decides to exit their staking positions from a Manual or Auto-Compounded Reward Pool, they have the power to initiate an undelegation. This process, similar to when they initially entered the Joining Pool, is a two-step journey. The user signs an intent to remove a specific delegation and patiently waits for at least one entire session before the operation can be executed by anyone.

Upon executing the leaving transaction intent, the protocol exchanges shares of the specified pool for native tokens at the current pool price. For the Manual Rewards Pool, any unclaimed rewards are assigned to the user. Simultaneously, the protocol purchases Leaving Pool shares in a one-to-one ratio for the native tokens the user just received. This ensures that the user joins the Leaving Pool, acquiring shares that correspond to the number of native tokens they desire to unstake.

After an entire session passes, any user can execute the pending operation. Then, the protocol swaps Leaving Pool shares for native protocol tokens at a one-to-one ratio.

> The primary purpose of the Leaving Pool is to provide a buffer for users leaving the staking mechanics. Consequently, through this buffer, the protocol can implement a slashing mechanism that affects undelegating users, but their stake was active during the slashable event. Nevertheless, slashing has not been implemented in Tanssi.

Reviewer: This paragraph is a bit confusing at the end, mainly this part: 

> the protocol can implement a slashing mechanism that affects undelegating users, but their stake was active during the slashable event.

 I would simplify to slashing has not been implemented in Tanssi but could be implemented in the future.

Author: What about:

```
The primary purpose of the Leaving Pool is to provide a buffer for users leaving the staking mechanics. This buffer allows to implement slashing mechanisms to deter bad behavior, but this has not been implemented yet.
```

The Leaving Pool's only purpose is for slashing haha

![High-level overview of the Leaving Pool when Staking](/images/learn/tanssi/staking/staking-5.webp)

### Swapping Between Rewards Pools {: #swap-rewards-pool}

Tanssi's staking module allows users to swap their stake from one type of reward pool to another. Users can use this functionality to move partial or full amounts of the staked tokens in a specific pool. The main benefit is that users don't have to go through the Leaving Pool and the Joining Pool again to move their stake.

First, all pending Manual Rewards Pool rewards are claimed at a protocol level, as liquidity is either added or removed. Therefore, the checkpoint rewards counter needs to be synced with the pool. Next, shares from the original pool are consumed and exchanged in favor of native protocol tokens at the current pool price. Then, shares of the new pool are attained at that pool's price. Lastly, any dust tokens remaining are automatically exchanged in favor of Leaving Pool shares. Note that all of the above is executed in the same block, and users don't have to wait for delays to earn rewards in the new pool. The dust in the Leaving Pool can be claimed after the required delays have passed.

> ![High-level overview of Swapping between Manual and AUto-Compounded Pools when Staking](/images/learn/tanssi/staking/staking-6.webp)

Reviewer: ```suggestion
![High-level overview of Swapping between Manual and Auto-Compounded Pools when Staking](/images/learn/tanssi/staking/staking-6.webp)
```
